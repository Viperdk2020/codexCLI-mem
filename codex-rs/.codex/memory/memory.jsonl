{"id":"92f4d16c-40c8-4821-b4c3-61bbce5b103a","created_at":"2025-09-02T03:12:33.353882193+00:00","updated_at":"2025-09-02T03:12:33.353882193+00:00","schema_version":1,"source":"codex-gui","scope":"Repo","status":"Active","kind":"Note","content":"on WLS","tags":[],"relevance_hints":{"files":[],"crates":[],"languages":[],"commands":[]},"counters":{"seen_count":0,"used_count":2,"last_used_at":"2025-09-02T14:04:03.636592338+00:00"},"expiry":null}
{"id":"ccdc0676-3830-4ab1-95ce-9b7d9cdf0fa1","created_at":"2025-09-02T12:39:46.351569665+00:00","updated_at":"2025-09-02T12:39:46.351569665+00:00","schema_version":1,"source":"codex-gui(headless)","scope":"Repo","status":"Active","kind":"Note","content":"Demo note: GUI headless save at 2025-09-02T12:39:45Z","tags":[],"relevance_hints":{"files":[],"crates":[],"languages":[],"commands":[]},"counters":{"seen_count":0,"used_count":1,"last_used_at":"2025-09-02T14:04:03.636592338+00:00"},"expiry":null}
{"id":"44705119-c86a-4689-bd4c-390c32a1dbe6","created_at":"2025-09-02T13:38:40.986968549+00:00","updated_at":"2025-09-02T13:38:40.986968549+00:00","schema_version":1,"source":"codex-gui","scope":"Repo","status":"Active","kind":"Note","content":"testing memory","tags":[],"relevance_hints":{"files":[],"crates":[],"languages":[],"commands":[]},"counters":{"seen_count":0,"used_count":1,"last_used_at":"2025-09-02T14:04:03.636592338+00:00"},"expiry":null}
{"id":"318f4941-938a-4942-9b84-1d3df5e4d1ac","created_at":"2025-09-02T14:04:03.221119247+00:00","updated_at":"2025-09-02T14:04:03.221119247+00:00","schema_version":1,"source":"codex-gui(headless)","scope":"Repo","status":"Active","kind":"Note","content":"use rg for search","tags":[],"relevance_hints":{"files":[],"crates":[],"languages":[],"commands":[]},"counters":{"seen_count":0,"used_count":1,"last_used_at":"2025-09-02T14:04:03.636592338+00:00"},"expiry":null}
