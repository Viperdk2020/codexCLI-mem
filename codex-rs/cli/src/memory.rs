use chrono::Utc;
use clap::Parser;
use codex_memory::factory;
use codex_memory::types::Counters;
use codex_memory::types::Kind;
use codex_memory::types::MemoryItem;
use codex_memory::types::RelevanceHints;
use codex_memory::types::Scope;
use codex_memory::types::Status;
use std::path::PathBuf;
use uuid::Uuid;

/// CLI for memory management commands.
#[derive(Debug, Parser)]
pub struct MemoryCli {
    #[command(subcommand)]
    pub cmd: MemoryCommand,
}

/// Memory subcommands.
#[derive(Debug, clap::Subcommand)]
pub enum MemoryCommand {
    /// Add a new memory item with given content.
    Add { content: String },
    /// List memory items.
    List,
    /// Edit an existing memory item.
    Edit { id: String, content: String },
    /// Remove a memory item by id.
    Rm { id: String },
    /// Archive a memory item.
    Archive { id: String },
    /// Unarchive a memory item.
    Unarchive { id: String },
    /// Export memory items to stdout.
    Export,
    /// Import memory items from stdin.
    Import,
    /// Migrate a JSONL file to a SQLite database.
    Migrate {
        /// Path to the source JSONL file
        #[arg(long)]
        jsonl: PathBuf,
        /// Path to the destination SQLite database file
        #[arg(long)]
        sqlite: PathBuf,
    },
    /// Compact a JSONL file by removing duplicate entries.
    Compact {
        /// Input JSONL file to compact
        #[arg(long)]
        input: PathBuf,
        /// Output JSONL file to write results
        #[arg(long)]
        output: PathBuf,
    },
    /// Show basic statistics about stored memories.
    Stats,
    /// Recall memories for a given prompt.
    Recall {
        #[arg(long = "for")]
        query: String,
    },
}

/// Execute the memory command.
pub fn run(cli: MemoryCli) -> anyhow::Result<()> {
    match cli.cmd {
        MemoryCommand::Migrate { jsonl, sqlite } => {
            let n = codex_memory::migrate::migrate_jsonl_to_sqlite(&jsonl, &sqlite)?;
            println!("Migrated {n} entries");
        }
        cmd => {
            let repo_root = std::env::current_dir()?;
            let store = factory::open_repo_store(&repo_root, None)?;
            match cmd {
                MemoryCommand::Add { content } => {
                    let now = Utc::now().to_rfc3339();
                    let item = MemoryItem {
                        id: Uuid::new_v4().to_string(),
                        created_at: now.clone(),
                        updated_at: now,
                        schema_version: 1,
                        source: "codex-cli".into(),
                        scope: Scope::Repo,
                        status: Status::Active,
                        kind: Kind::Note,
                        content,
                        tags: Vec::new(),
                        relevance_hints: RelevanceHints {
                            files: Vec::new(),
                            crates: Vec::new(),
                            languages: Vec::new(),
                            commands: Vec::new(),
                        },
                        counters: Counters {
                            seen_count: 0,
                            used_count: 0,
                            last_used_at: None,
                        },
                        expiry: None,
                    };
                    store.add(item)?;
                }
                MemoryCommand::List => {
                    for item in store.list(None, None)? {
                        println!("{}", item.content);
                    }
                }
                MemoryCommand::Edit { id, content } => {
                    if let Some(mut item) = store.get(&id)? {
                        item.content = content;
                        item.updated_at = Utc::now().to_rfc3339();
                        store.update(&item)?;
                    } else {
                        anyhow::bail!("memory id not found: {id}");
                    }
                }
                MemoryCommand::Rm { id } => {
                    store.delete(&id)?;
                }
                MemoryCommand::Archive { id } => {
                    store.archive(&id, true)?;
                }
                MemoryCommand::Unarchive { id } => {
                    store.archive(&id, false)?;
                }
                MemoryCommand::Export => {
                    let mut out = std::io::stdout();
                    store.export(&mut out)?;
                }
                MemoryCommand::Import => {
                    let mut input = std::io::stdin();
                    let n = store.import(&mut input)?;
                    println!("Imported {n} items");
                }
                MemoryCommand::Stats => {
                    let stats = store.stats()?;
                    println!("{stats}");
                }
                MemoryCommand::Recall { query } => {
                    let ctx = codex_memory::recall::RecallContext {
                        repo_root: Some(repo_root),
                        dir: None,
                        current_file: None,
                        crate_name: None,
                        language: None,
                        command: None,
                        now_rfc3339: Utc::now().to_rfc3339(),
                        item_cap: 8,
                        token_cap: 300,
                    };
                    let items = codex_memory::recall::recall(store.as_ref(), &query, &ctx)?;
                    println!("{}", serde_json::to_string(&items)?);
                }
                MemoryCommand::Compact { input, output } => {
                    let (read, written) = codex_memory::migrate::compact_jsonl(&input, &output)?;
                    println!("Read {read} entries, wrote {written} entries");
                }
                MemoryCommand::Migrate { .. } => unreachable!(),
            }
        }
    }
    Ok(())
}
