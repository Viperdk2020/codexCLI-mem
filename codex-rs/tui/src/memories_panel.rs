use chrono::Utc;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::text::Line;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Widget;
use ratatui::widgets::WidgetRef;
use uuid::Uuid;

use codex_memory::factory;
use codex_memory::types::Counters;
use codex_memory::types::Kind;
use codex_memory::types::MemoryItem;
use codex_memory::types::RelevanceHints;
use codex_memory::types::Scope;
use codex_memory::types::Status;

/// Simple panel showing stored memories and exposing minimal CRUD ops.
pub struct MemoriesPanel {
    repo_root: std::path::PathBuf,
    items: Vec<MemoryItem>,
}

impl MemoriesPanel {
    pub fn new(repo_root: std::path::PathBuf) -> anyhow::Result<Self> {
        let mut panel = Self {
            repo_root,
            items: Vec::new(),
        };
        panel.refresh()?;
        Ok(panel)
    }

    /// Reload items from the repo store.
    pub fn refresh(&mut self) -> anyhow::Result<()> {
        let store = factory::open_repo_store(&self.repo_root, None)?;
        self.items = store.list(Some(Scope::Repo), Some(Status::Active))?;
        Ok(())
    }

    /// Add a new preference memory entry.
    pub fn add_pref(&mut self, text: &str) -> anyhow::Result<()> {
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let item = MemoryItem {
            id: Uuid::new_v4().to_string(),
            created_at: ts.clone(),
            updated_at: ts,
            schema_version: 1,
            source: "codex-tui".to_string(),
            scope: Scope::Repo,
            status: Status::Active,
            kind: Kind::Pref,
            content: text.to_string(),
            tags: vec!["pref".to_string()],
            relevance_hints: RelevanceHints {
                files: vec![],
                crates: vec![],
                languages: vec![],
                commands: vec![],
            },
            counters: Counters {
                seen_count: 0,
                used_count: 0,
                last_used_at: None,
            },
            expiry: None,
        };
        let store = factory::open_repo_store(&self.repo_root, None)?;
        store.add(item)?;
        self.refresh()
    }

    /// Delete a memory item by id.
    pub fn delete(&mut self, id: &str) -> anyhow::Result<()> {
        let store = factory::open_repo_store(&self.repo_root, None)?;
        store.delete(id)?;
        self.refresh()
    }
}

impl WidgetRef for &MemoriesPanel {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        let mut lines: Vec<Line<'static>> = Vec::with_capacity(self.items.len() + 1);
        lines.push(Line::raw("Memories:"));
        for it in &self.items {
            lines.push(Line::raw(format!("- {}", it.content)));
        }
        let para = Paragraph::new(lines);
        para.render(area, buf);
    }
}
