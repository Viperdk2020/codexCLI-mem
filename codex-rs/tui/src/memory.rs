use chrono::Utc;
use serde_json::json;
use std::fs::OpenOptions;
use std::fs::create_dir_all;
use std::io::BufRead;
use std::io::Write;
use std::path::Path;
use std::path::PathBuf;
use std::time::Duration;
use uuid::Uuid;

use codex_memory::factory;
use codex_memory::recall::RecallContext;
use codex_memory::recall::{self};
use codex_memory::types::Counters;
use codex_memory::types::Kind;
use codex_memory::types::MemoryItem;
use codex_memory::types::RelevanceHints;
use codex_memory::types::Scope;
use codex_memory::types::Status;

pub struct MemoryLogger {
    repo_root: PathBuf,
    memory_dir: PathBuf,
    memory_file: PathBuf,
    session_id: Option<String>,
}

#[derive(Debug, Clone)]
pub(crate) struct ToolInvocation {
    pub server: String,
    pub tool: String,
    pub arguments: Option<serde_json::Value>,
    pub duration: Duration,
    pub success: bool,
    pub result: Option<serde_json::Value>,
}

impl MemoryLogger {
    pub fn new(start_path: PathBuf) -> Self {
        let repo_root = detect_repo_root(&start_path).unwrap_or(start_path);
        let memory_dir = repo_root.join(".codex").join("memory");
        let memory_file = memory_dir.join("memory.jsonl");
        let _ = create_dir_all(&memory_dir);
        Self {
            repo_root,
            memory_dir,
            memory_file,
            session_id: None,
        }
    }

    pub fn set_session_id(&mut self, session_id: Uuid) {
        self.session_id = Some(session_id.to_string());
    }

    fn write_line(&self, value: &serde_json::Value) {
        if let Err(e) = create_dir_all(&self.memory_dir) {
            tracing::debug!("tui memory: create_dir_all failed: {e}");
            return;
        }
        match OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.memory_file)
        {
            Ok(mut f) => {
                if let Ok(s) = serde_json::to_string(value) {
                    let _ = writeln!(f, "{s}");
                }
            }
            Err(e) => tracing::debug!("tui memory: open append failed: {e}"),
        }
    }

    pub fn log_exec(&self, command: &[String], exit_code: i32, duration: Duration, output: &str) {
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let preview = truncate_multiline(output, 160, 20);
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "exec",
            "content": shlex::try_join(command.iter().map(|s| s.as_str())).unwrap_or_else(|_| command.join(" ")),
            "tags": ["exec"],
            "files": [],
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {
                "exit_code": exit_code,
                "duration_ms": duration.as_millis() as u64,
                "output_preview": preview,
            }
        });
        self.write_line(&value);
    }

    pub fn log_tool_call(&self, inv: ToolInvocation) {
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let args_str = inv
            .arguments
            .as_ref()
            .map(|v| serde_json::to_string(v).unwrap_or_default())
            .unwrap_or_default();
        let content = if args_str.is_empty() {
            format!("{}.{}()", inv.server, inv.tool)
        } else {
            format!("{}.{}({})", inv.server, inv.tool, args_str)
        };
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "tool",
            "content": content,
            "tags": ["tool"],
            "files": [],
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {
                "server": inv.server,
                "tool": inv.tool,
                "success": inv.success,
                "duration_ms": inv.duration.as_millis() as u64,
                "result": inv.result,
            }
        });
        self.write_line(&value);
    }

    pub fn log_patch_apply(
        &self,
        success: bool,
        auto_approved: bool,
        duration: Duration,
        stdout: &str,
        stderr: &str,
        files: &[String],
    ) {
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let preview = if success { stdout } else { stderr };
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "change",
            "content": format!("apply_patch(auto_approved={})", auto_approved),
            "tags": ["apply_patch"],
            "files": files,
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {
                "success": success,
                "auto_approved": auto_approved,
                "duration_ms": duration.as_millis() as u64,
                "output_preview": truncate_multiline(preview, 160, 20),
            }
        });
        self.write_line(&value);
    }

    // --- Durable items API ---
    pub fn add_summary(&self, text: &str) -> anyhow::Result<()> {
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "summary",
            "content": text,
            "tags": ["summary"],
            "files": [],
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {}
        });
        self.write_line(&value);
        Ok(())
    }

    pub fn log_decision(&self, text: &str, tags: &[&str]) {
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "decision",
            "content": text,
            "tags": tags,
            "files": [],
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {}
        });
        self.write_line(&value);
    }

    // Build a short preamble string from durable memory items (prefs/summaries).
    pub fn build_durable_preamble(&self, max_len: usize) -> Option<String> {
        let store = factory::open_repo_store(&self.repo_root, None).ok()?;
        let ctx = RecallContext {
            repo_root: Some(self.repo_root.clone()),
            dir: None,
            current_file: None,
            crate_name: None,
            language: None,
            command: None,
            now_rfc3339: Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
            item_cap: 16,
            token_cap: max_len * 2,
        };
        let Ok(items) = recall::recall(store.as_ref(), "", &ctx) else {
            return None;
        };
        let mut prefs: Vec<(String, Vec<String>)> = Vec::new();
        let mut summaries: Vec<(String, Vec<String>)> = Vec::new();
        for it in items {
            match it.kind {
                Kind::Pref => prefs.push((it.content, it.tags)),
                Kind::Fact => summaries.push((it.content, it.tags)),
                _ => {}
            }
        }
        if prefs.is_empty() && summaries.is_empty() {
            return None;
        }
        let dedupe = |items: Vec<(String, Vec<String>)>, cap: usize| -> Vec<String> {
            use std::collections::BTreeMap;
            let mut map: BTreeMap<String, (Vec<String>, usize)> = BTreeMap::new();
            for (c, tags) in items {
                let key = c.to_ascii_lowercase();
                let e = map.entry(key).or_insert((Vec::new(), 0));
                for t in tags {
                    if !e.0.contains(&t) {
                        e.0.push(t);
                    }
                }
                e.1 += 1;
            }
            let mut out: Vec<String> = map
                .into_iter()
                .map(|(k, (tags, cnt))| {
                    if cnt > 1 && !tags.is_empty() {
                        format!("{k} (tags: {} ×{cnt})", tags.join(", "))
                    } else if !tags.is_empty() {
                        format!("{k} (tags: {})", tags.join(", "))
                    } else {
                        k
                    }
                })
                .collect();
            if out.len() > cap {
                out.truncate(cap);
            }
            out
        };
        let prefs_out = dedupe(prefs, 8);
        let summaries_out = dedupe(summaries, 6);
        let mut parts: Vec<String> = Vec::new();
        if !prefs_out.is_empty() {
            parts.push(format!(
                "Project preferences:\n- {}",
                prefs_out.join("\n- ")
            ));
        }
        if !summaries_out.is_empty() {
            parts.push(format!("Project facts:\n- {}", summaries_out.join("\n- ")));
        }
        let mut s = parts.join("\n\n");
        if s.len() > max_len {
            s.truncate(max_len);
            s.push_str("\n…");
        }
        Some(format!(
            "Context: The following project memory may be helpful.\n{s}\nPlease follow these preferences and consider these facts."
        ))
    }

    // Minimal durable ops for TUI slash commands.
    pub fn add_pref(&self, text: &str) -> anyhow::Result<()> {
        if sqlite_enabled() {
            #[cfg(feature = "memory-sqlite")]
            {
                let id = Uuid::new_v4().to_string();
                let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
                let item = MemoryItem {
                    id,
                    created_at: ts.clone(),
                    updated_at: ts,
                    schema_version: 1,
                    source: "codex-tui".to_string(),
                    scope: Scope::Repo,
                    status: Status::Active,
                    kind: Kind::Pref,
                    content: text.to_string(),
                    tags: vec!["pref".to_string()],
                    relevance_hints: RelevanceHints {
                        files: vec![],
                        crates: vec![],
                        languages: vec![],
                        commands: vec![],
                    },
                    counters: Counters {
                        seen_count: 0,
                        used_count: 0,
                        last_used_at: None,
                    },
                    expiry: None,
                };
                let store = factory::open_repo_store(&self.repo_root, None)?;
                store.add(item)?;
                return Ok(());
            }
        }
        let id = Uuid::new_v4().to_string();
        let ts = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let value = json!({
            "id": id,
            "ts": ts,
            "repo": self.repo_root.to_string_lossy(),
            "type": "pref",
            "content": text,
            "tags": ["pref"],
            "files": [],
            "session_id": self.session_id,
            "source": "codex-tui",
            "metadata": {}
        });
        self.write_line(&value);
        Ok(())
    }

    pub fn list_durable(&self, limit: usize) -> Vec<DurableItem> {
        if sqlite_enabled() {
            #[cfg(feature = "memory-sqlite")]
            {
                let Ok(store) = factory::open_repo_store(&self.repo_root, None) else {
                    return vec![];
                };
                let Ok(mut items) = store.list(Some(Scope::Repo), Some(Status::Active)) else {
                    return vec![];
                };
                // Only preferences and summaries (facts)
                items.retain(|i| matches!(i.kind, Kind::Pref | Kind::Fact));
                items.truncate(limit);
                return items
                    .into_iter()
                    .map(|i| DurableItem {
                        id: i.id,
                        r#type: match i.kind {
                            Kind::Pref => "pref".to_string(),
                            _ => "summary".to_string(),
                        },
                        content: i.content,
                    })
                    .collect();
            }
        }
        let Ok(file) = std::fs::File::open(&self.memory_file) else {
            return vec![];
        };
        let reader = std::io::BufReader::new(file);
        let mut items: Vec<DurableItem> = reader
            .lines()
            .flatten()
            .filter_map(|line| serde_json::from_str::<serde_json::Value>(&line).ok())
            .filter_map(|v| DurableItem::from_value(&v))
            .collect();
        if items.len() > limit {
            items.drain(0..items.len().saturating_sub(limit));
        }
        items
    }

    pub fn search_durable(&self, query: &str, limit: usize) -> Vec<DurableItem> {
        if sqlite_enabled() {
            #[cfg(feature = "memory-sqlite")]
            {
                let q = query.to_ascii_lowercase();
                return self
                    .list_durable(usize::MAX)
                    .into_iter()
                    .filter(|i| i.content.to_ascii_lowercase().contains(&q))
                    .take(limit)
                    .collect();
            }
        }
        let q = query.to_ascii_lowercase();
        self.list_durable(usize::MAX)
            .into_iter()
            .filter(|i| i.content.to_ascii_lowercase().contains(&q))
            .take(limit)
            .collect()
    }

    pub fn list_durable_tagged(&self, limit: usize, tag: &str) -> Vec<DurableItem> {
        if sqlite_enabled() {
            #[cfg(feature = "memory-sqlite")]
            {
                let t = tag.to_ascii_lowercase();
                let Ok(store) = factory::open_repo_store(&self.repo_root, None) else {
                    return vec![];
                };
                let Ok(mut items) = store.list(Some(Scope::Repo), Some(Status::Active)) else {
                    return vec![];
                };
                items.retain(|i| {
                    i.tags.iter().any(|x| x.eq_ignore_ascii_case(&t))
                        && matches!(i.kind, Kind::Pref | Kind::Fact)
                });
                items.truncate(limit);
                return items
                    .into_iter()
                    .map(|i| DurableItem {
                        id: i.id,
                        r#type: match i.kind {
                            Kind::Pref => "pref".to_string(),
                            _ => "summary".to_string(),
                        },
                        content: i.content,
                    })
                    .collect();
            }
        }
        let t = tag.to_ascii_lowercase();
        let Ok(file) = std::fs::File::open(&self.memory_file) else {
            return vec![];
        };
        let reader = std::io::BufReader::new(file);
        let mut items: Vec<DurableItem> = reader
            .lines()
            .flatten()
            .filter_map(|line| serde_json::from_str::<serde_json::Value>(&line).ok())
            .filter_map(|v| {
                let has_tag = v
                    .get("tags")
                    .and_then(|x| x.as_array())
                    .map(|arr| {
                        arr.iter().any(|t0| {
                            t0.as_str()
                                .map(|s| s.eq_ignore_ascii_case(&t))
                                .unwrap_or(false)
                        })
                    })
                    .unwrap_or(false);
                if has_tag {
                    DurableItem::from_value(&v)
                } else {
                    None
                }
            })
            .collect();
        if items.len() > limit {
            items.drain(0..items.len().saturating_sub(limit));
        }
        items
    }

    pub fn delete_by_prefix(&self, prefix: &str) -> bool {
        if sqlite_enabled() {
            #[cfg(feature = "memory-sqlite")]
            {
                if prefix.is_empty() {
                    return false;
                }
                if let Ok(store) = factory::open_repo_store(&self.repo_root, None)
                    && let Ok(items) = store.list(Some(Scope::Repo), None)
                {
                    let mut changed = false;
                    for it in items {
                        if (matches!(it.kind, Kind::Pref | Kind::Fact)) && it.id.starts_with(prefix)
                        {
                            let _ = store.delete(&it.id);
                            changed = true;
                        }
                    }
                    return changed;
                }
                return false;
            }
        }
        if prefix.is_empty() {
            return false;
        }
        let Ok(file) = std::fs::File::open(&self.memory_file) else {
            return false;
        };
        let reader = std::io::BufReader::new(&file);
        let lines: Vec<String> = reader.lines().flatten().collect();
        let mut changed = false;
        let mut out: Vec<String> = Vec::with_capacity(lines.len());
        for line in lines {
            if let Ok(v) = serde_json::from_str::<serde_json::Value>(&line) {
                let t = v.get("type").and_then(|x| x.as_str()).unwrap_or("");
                let is_durable = t == "pref" || t == "summary";
                let id = v.get("id").and_then(|x| x.as_str()).unwrap_or("");
                if is_durable && id.starts_with(prefix) {
                    changed = true;
                    continue;
                }
            }
            out.push(line);
        }
        if changed
            && let Ok(mut f) = std::fs::OpenOptions::new()
                .write(true)
                .truncate(true)
                .open(&self.memory_file)
        {
            let _ = writeln!(f, "{}", out.join("\n"));
        }
        changed
    }
}

#[derive(Clone)]
pub(crate) struct DurableItem {
    pub id: String,
    pub r#type: String,
    pub content: String,
}
impl DurableItem {
    fn from_value(v: &serde_json::Value) -> Option<Self> {
        let t = v.get("type")?.as_str()?;
        if t != "pref" && t != "summary" {
            return None;
        }
        Some(DurableItem {
            id: v.get("id")?.as_str()?.to_string(),
            r#type: t.to_string(),
            content: v.get("content")?.as_str()?.to_string(),
        })
    }
}

fn truncate_multiline(text: &str, max_chars: usize, max_lines: usize) -> String {
    let mut s: String = text.lines().take(max_lines).collect::<Vec<_>>().join("\n");
    if s.len() > max_chars {
        s.truncate(max_chars);
        s.push('…');
    }
    s
}

fn detect_repo_root(start: &Path) -> Option<PathBuf> {
    let mut cur = start.canonicalize().unwrap_or(start.to_path_buf());
    for _ in 0..64 {
        if cur.join(".git").exists() || cur.join(".codex").exists() {
            return Some(cur);
        }
        if let Some(parent) = cur.parent() {
            cur = parent.to_path_buf();
        } else {
            break;
        }
    }
    None
}

fn sqlite_enabled() -> bool {
    #[cfg(feature = "memory-sqlite")]
    {
        match std::env::var("CODEX_MEMORY_BACKEND") {
            Ok(v) if v.eq_ignore_ascii_case("sqlite") => return true,
            _ => {}
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::tempdir;

    fn write_jsonl<P: AsRef<Path>>(path: P, lines: &[serde_json::Value]) {
        let mut s = String::new();
        for v in lines {
            s.push_str(&serde_json::to_string(v).unwrap());
            s.push('\n');
        }
        fs::write(path, s).unwrap();
    }

    #[test]
    fn preamble_dedupes_merges_and_caps() {
        let dir = tempdir().unwrap();
        let repo = dir.path().to_path_buf();
        fs::create_dir_all(repo.join(".codex").join("memory")).unwrap();
        let store = factory::open_repo_store(&repo, None).unwrap();
        let ts = "2025-01-01T00:00:00.000Z".to_string();
        let item = MemoryItem {
            id: "1".into(),
            created_at: ts.clone(),
            updated_at: ts.clone(),
            schema_version: 1,
            source: "test".into(),
            scope: Scope::Repo,
            status: Status::Active,
            kind: Kind::Pref,
            content: "prefer ruff".into(),
            tags: vec!["python".into(), "style".into()],
            relevance_hints: RelevanceHints { files: vec![], crates: vec![], languages: vec![], commands: vec![] },
            counters: Counters { seen_count: 0, used_count: 0, last_used_at: None },
            expiry: None,
        };
        store.add(item).unwrap();
        let mut item2 = store.get("1").unwrap().unwrap();
        item2.id = "2".into();
        item2.content = "Prefer Ruff".into();
        item2.tags = vec!["style".into()];
        store.add(item2).unwrap();
        let mut fact = store.get("1").unwrap().unwrap();
        fact.id = "3".into();
        fact.kind = Kind::Fact;
        fact.content = "uses pytest".into();
        fact.tags = vec!["python".into()];
        store.add(fact).unwrap();
        let mut fact2 = store.get("3").unwrap().unwrap();
        fact2.id = "4".into();
        fact2.content = "Uses PyTest".into();
        fact2.tags = vec!["ci".into()];
        store.add(fact2).unwrap();

        let mut logger = MemoryLogger::new(repo.clone());
        logger.session_id = Some("test".into());
        let pre = logger.build_durable_preamble(512).expect("preamble");

        // Expect deduped entries each appearing once and mention of tags.
        assert!(pre.to_ascii_lowercase().contains("project preferences"));
        assert!(pre.to_ascii_lowercase().contains("project facts"));
        // The two Ruff prefs should merge into one line (case-insensitive dedupe) and include tags
        assert!(pre.to_ascii_lowercase().contains("prefer ruff"));
        // The two pytest summaries should merge
        assert!(pre.to_ascii_lowercase().contains("uses pytest"));
    }
}
