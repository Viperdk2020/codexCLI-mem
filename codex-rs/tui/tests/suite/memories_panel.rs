use codex_tui::memories_panel::MemoriesPanel;
use codex_tui::memory::MemoryLogger;
use ratatui::Terminal;
use ratatui::backend::TestBackend;
use ratatui::widgets::WidgetRef;
use tempfile::tempdir;

#[test]
fn panel_renders() {
    let dir = tempdir().unwrap();
    std::fs::create_dir(dir.path().join(".codex")).unwrap();
    let mut panel = MemoriesPanel::new(dir.path().to_path_buf()).unwrap();
    panel.add_pref("Use ripgrep for search").unwrap();
    panel.add_pref("Avoid force pushes").unwrap();
    panel.refresh().unwrap();

    let mut terminal = Terminal::new(TestBackend::new(40, 6)).unwrap();
    terminal
        .draw(|f| (&panel).render_ref(f.area(), f.buffer_mut()))
        .unwrap();
    insta::assert_snapshot!(terminal.backend());
}

#[test]
fn preamble_preview() {
    let dir = tempdir().unwrap();
    std::fs::create_dir(dir.path().join(".codex")).unwrap();
    let repo = dir.path().to_path_buf();
    let store = codex_memory::factory::open_repo_store(&repo, None).unwrap();
    let ts = "2025-01-01T00:00:00.000Z".to_string();
    let pref = codex_memory::types::MemoryItem {
        id: "1".into(),
        created_at: ts.clone(),
        updated_at: ts.clone(),
        schema_version: 1,
        source: "test".into(),
        scope: codex_memory::types::Scope::Repo,
        status: codex_memory::types::Status::Active,
        kind: codex_memory::types::Kind::Pref,
        content: "Respect editorconfig".into(),
        tags: vec![],
        relevance_hints: codex_memory::types::RelevanceHints {
            files: vec![],
            crates: vec![],
            languages: vec![],
            commands: vec![],
        },
        counters: codex_memory::types::Counters {
            seen_count: 0,
            used_count: 0,
            last_used_at: None,
        },
        expiry: None,
    };
    store.add(pref).unwrap();
    let mut fact = store.get("1").unwrap().unwrap();
    fact.id = "2".into();
    fact.kind = codex_memory::types::Kind::Fact;
    fact.content = "Tests use cargo test".into();
    store.add(fact).unwrap();
    let logger = MemoryLogger::new(repo);
    let pre = logger.build_durable_preamble(512).unwrap();
    insta::assert_snapshot!(pre);
}
