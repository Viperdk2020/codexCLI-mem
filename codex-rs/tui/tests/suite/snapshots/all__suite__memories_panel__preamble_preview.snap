---
source: tui/tests/suite/memories_panel.rs
assertion_line: 54
expression: pre
---
Context: The following project memory may be helpful.
Project preferences:
- respect editorconfig

Project facts:
- tests use cargo test
Please follow these preferences and consider these facts.
